src/SignalR.Orleans/OrleansHubLifetimeManager.cs

```diff
@@ -151,8 +145,12 @@ public override Task InvokeGroupAsync(string groupName, string methodName, objec
 
         public override Task InvokeUserAsync(string userId, string methodName, object[] args)
         {
-            // TODO: Check with @davidfowl what is this method for
-            throw new NotImplementedException();
+            if (string.IsNullOrWhiteSpace(userId)) throw new ArgumentNullException(nameof(userId));
+            if (string.IsNullOrWhiteSpace(methodName)) throw new ArgumentNullException(nameof(methodName));
+
+            var message = new InvocationMessage(Guid.NewGuid().ToString(), nonBlocking: true, target: methodName, arguments: args);
+            var grain = this._clusterClient.GetGrain<IClientGrain>(userId);

```

Comment:
I don't think this is correct... `IClientGrain` expect the ID to be the ConnectionId, not the userId.

Reply:
I'm not sure but I think these are the same. This worked fine on my sample application.

[Here I m setting the user ConnectionId](
https://github.com/claylaut/SignalR.Orleans/blob/feature/first-stable-messaging/sample/Sample.Orleans.Client.WebApp/SignalRHubs/HeroHub.cs#L38) and the next line I sending this specific message only to this particular user.

Reply:
yeah but you can get `Clients.GetClient(Context.ConnectionId)` as well. They are all strings. I'm just concerned that the semantics is not right and that a `User` means something different from Client/Connection in the context of SignalR.

Reply:
You have a valid point, better double check

Reply:
For me, it looks like User is just a different way to address a client across multiple connections if I understood correctly. If we look at the [Redis implementation](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs#L533-L552) we will find that it is just a subscription that is created [IF the client has a particular UserIdentifier](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs#L110-L113) set. If that is the case, we should have another grain called `IUserGrain` almost the same as we have the ClientGrain and we should make the relation between them to work by using streams. 

Again, this is my understanding of it. The code in Redis sample is more complicate then ours, since it has intrinsics of Redis like scripts etc. We are mostly relying in simple grain calls and few SMS streams. I would really appreciate @davidfowl suggestion on that.

Reply:
Good catch, after re checked the code, this implementation is wrong. 

I think the difference between `UserId` and `ConnectionId`  is a mapping of `User` with one or multiple connections. (for example multiple browser tabs)

https://docs.microsoft.com/en-us/aspnet/signalr/overview/guide-to-the-api/mapping-users-to-connections#IUserIdProvider

Reply:
Ok! Thats is a good explanation. I'll have it implemented asap. If you can submit a PR just with the change on that wrong IF statement I would appreciate that.

Thanks!

Reply:
I think we can use the `GroupGrain` since its exactly the same. But I need to implement a sample of Authentication to be able to test it out properly.

Something to keep in mind ideally, we should use the [connection.UserIdentifier](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Core/DefaultHubLifetimeManager.cs#L134) but currently, it's not available with the published version.

I will open a pr for the `if` statement.